File: server/internal/models/voiceprint.go
package models

import (
	"time"
)

// Voiceprint 声纹记录模型
type Voiceprint struct {
	ID            int       `json:"id" gorm:"primaryKey;autoIncrement"`
	SpeakerID     string    `json:"speaker_id" gorm:"type:varchar(255);not null"`
	AssistantID   string    `json:"assistant_id" gorm:"type:varchar(255);not null"`
	SpeakerName   string    `json:"speaker_name" gorm:"type:varchar(255)"`
	FeatureVector []byte    `json:"-" gorm:"type:longblob;not null"`
	CreatedAt     time.Time `json:"created_at" gorm:"autoCreateTime"`
	UpdatedAt     time.Time `json:"updated_at" gorm:"autoUpdateTime"`
}

// TableName 指定表名
func (Voiceprint) TableName() string {
	return "voiceprints"
}

// VoiceprintCreateRequest 创建声纹请求
type VoiceprintCreateRequest struct {
	SpeakerID   string `json:"speaker_id" binding:"required"`
	AssistantID string `json:"assistant_id" binding:"required"`
	SpeakerName string `json:"speaker_name" binding:"required"`
}

// VoiceprintUpdateRequest 更新声纹请求
type VoiceprintUpdateRequest struct {
	SpeakerName string `json:"speaker_name"`
}

// VoiceprintResponse 声纹响应
type VoiceprintResponse struct {
	ID          int        `json:"id"`
	SpeakerID   string     `json:"speaker_id"`
	AssistantID string     `json:"assistant_id"`
	SpeakerName string     `json:"speaker_name"`
	CreatedAt   time.Time  `json:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at"`
	LastUsed    *time.Time `json:"last_used,omitempty"`
	Confidence  float64    `json:"confidence,omitempty"`
}

// VoiceprintListResponse 声纹列表响应
type VoiceprintListResponse struct {
	Total       int                  `json:"total"`
	Voiceprints []VoiceprintResponse `json:"voiceprints"`
}

// VoiceprintRegisterRequest 声纹注册请求（用于调用voiceprint-api）
type VoiceprintRegisterRequest struct {
	SpeakerID   string `json:"speaker_id"`
	AssistantID string `json:"assistant_id"`
	AudioFile   []byte `json:"-"`
}

// VoiceprintIdentifyRequest 声纹识别请求
type VoiceprintIdentifyRequest struct {
	AssistantID  string   `json:"assistant_id" binding:"required"`
	CandidateIDs []string `json:"candidate_ids"`
	AudioFile    []byte   `json:"-"`
}

// VoiceprintIdentifyResponse 声纹识别响应
type VoiceprintIdentifyResponse struct {
	SpeakerID  string  `json:"speaker_id"`
	Score      float64 `json:"score"`
	Confidence string  `json:"confidence"`
	IsMatch    bool    `json:"is_match"`
}

// VoiceprintVerifyResponse 声纹验证响应
type VoiceprintVerifyResponse struct {
	TargetSpeakerID     string  `json:"target_speaker_id"`     // 目标说话人ID
	IdentifiedSpeakerID string  `json:"identified_speaker_id"` // 识别出的说话人ID
	Score               float64 `json:"score"`                 // 相似度分数
	Confidence          string  `json:"confidence"`            // 置信度等级
	IsMatch             bool    `json:"is_match"`              // 是否匹配（基于置信度阈值）
	IsTargetSpeaker     bool    `json:"is_target_speaker"`     // 是否为目标说话人
	VerificationPassed  bool    `json:"verification_passed"`   // 验证是否通过
}

File: web/src/api/voiceprint.ts
import { get, post, put, del } from '@/utils/request'
import { ApiResponse } from '@/utils/request'

// 声纹记录类型
export interface VoiceprintRecord {
  id: number
  speaker_id: string
  assistant_id: string
  speaker_name: string
  created_at: string
  updated_at: string
  last_used?: string
  confidence?: number
}

// 声纹列表响应
export interface VoiceprintListResponse {
  total: number
  voiceprints: VoiceprintRecord[]
}

// 创建声纹请求
export interface CreateVoiceprintRequest {
  speaker_id: string
  assistant_id: string
  speaker_name: string
}

// 更新声纹请求
export interface UpdateVoiceprintRequest {
  speaker_name?: string
}

// 声纹识别响应
export interface VoiceprintIdentifyResponse {
  speaker_id: string
  score: number
  confidence: string
  is_match: boolean
}

// 声纹验证响应
export interface VoiceprintVerifyResponse {
  target_speaker_id: string
  identified_speaker_id: string
  score: number
  confidence: string
  is_match: boolean
  is_target_speaker: boolean
  verification_passed: boolean
}

// 获取声纹列表
export const getVoiceprints = async (assistantId: string): Promise<ApiResponse<VoiceprintListResponse>> => {
  return get(`/voiceprint?assistant_id=${assistantId}`)
}

// 创建声纹记录
export const createVoiceprint = async (data: CreateVoiceprintRequest): Promise<ApiResponse<VoiceprintRecord>> => {
  return post('/voiceprint', data)
}

// 注册声纹（上传音频）
export const registerVoiceprint = async (
  assistantId: string,
  speakerName: string,
  audioFile: File
): Promise<ApiResponse<VoiceprintRecord>> => {
  const formData = new FormData()
  formData.append('assistant_id', assistantId)
  formData.append('speaker_name', speakerName)
  formData.append('audio_file', audioFile)

  return post('/voiceprint/register', formData)
}

// 更新声纹记录
export const updateVoiceprint = async (id: number, data: UpdateVoiceprintRequest): Promise<ApiResponse<VoiceprintRecord>> => {
  return put(`/voiceprint/${id}`, data)
}

// 删除声纹记录
export const deleteVoiceprint = async (id: number): Promise<ApiResponse<void>> => {
  return del(`/voiceprint/${id}`)
}

// 声纹识别
export const identifyVoiceprint = async (
  assistantId: string,
  audioFile: File
): Promise<ApiResponse<VoiceprintIdentifyResponse>> => {
  const formData = new FormData()
  formData.append('assistant_id', assistantId)
  formData.append('audio_file', audioFile)

  return post('/voiceprint/identify', formData)
}

// 声纹验证
export const verifyVoiceprint = async (
  assistantId: string,
  speakerId: string,
  audioFile: File
): Promise<ApiResponse<VoiceprintVerifyResponse>> => {
  const formData = new FormData()
  formData.append('assistant_id', assistantId)
  formData.append('speaker_id', speakerId)
  formData.append('audio_file', audioFile)

  return post('/voiceprint/verify', formData)
}